cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -527,6 +517,55 @@ struct Utf8CapitalizeTransform : public StringTransformBase {
 template <typename Type>
 using Utf8Capitalize = StringTransformExec<Type, Utf8CapitalizeTransform>;
 
+struct Utf8TitleTransform : public StringTransformCodepointBase {
+  int64_t Transform(const uint8_t* input, int64_t input_string_ncodeunits,
+                    uint8_t* output) {
+    uint8_t* output_start = output;
+    const uint8_t* end = input + input_string_ncodeunits;
+    const uint8_t* curr = NULLPTR;
+    uint32_t codepoint = 0;
+
+    do {
+      // Uppercase first alpha character of current word
+      while (input < end) {
+        curr = input;
+        if (ARROW_PREDICT_FALSE(!util::UTF8Decode(&curr, &codepoint))) {
+          return kTransformError;
+        }
+        if (IsCasedCharacterUnicode(codepoint)) {
+          output =
+              util::UTF8Encode(output, UTF8UpperTransform::TransformCodepoint(codepoint));
+          input = curr;
+          break;
+        }
+        output = std::copy(input, curr, output);
+        input = curr;
+      }
+
+      // Lowercase characters until a whitespace is found
```

Comment:
I don't think that's the expected semantics. In Python I get:
```python
>>> "foo bar;héhé0zop".title()
'Foo Bar;Héhé0Zop'
```

which means that any non-cased character should break this loop, not only whitespace characters.

Reply:
Thanks for catching this. I do have the incorrect semantics of what titlecase is. Lesson learned: Always verify multiple languages/sources for semantics of compute functions.

Reply:
[Python's `str.title`](https://docs.python.org/3/library/stdtypes.html#str.title) also discusses the need for handling special cases such as apostrophes.